### PR TITLE
Allow replication 'role reversal'

### PR DESCRIPTION
**Pull request by Vagabond**
In situations when you can only connect outbound because of firewall policies, etc, you can't do outbound replication because TCP connection direction and replication direction are tied together.

These changes allow you to subvert that by switching roles for a connection once a connection is established. However, these changes are kind of a hack and should only be used as a last resort.

Some caveats:
- Both sides need to be _configured_ to do this, it is not negotiated
- You MUST use the new 'keylist' fullsync strategy, at least if you want start/stop/pause/resume to work
- If you turn this on, it is NOT backwards compatible

### CODE REVIEW

File: src/riak_repl_tcp_client.erl [hunk 1]
```diff
@@ -68,6 +73,18 @@ init([SiteName]) ->
             {ok, do_async_connect(State)}
     end.
 
+handle_call(start_fullsync, _From, #state{fullsync_worker=FSW} = State) ->
```
Comment (jonmeredith) [inline src/riak_repl_tcp_client.erl:78]: Can we add a comment block to explain  why these got added here, they aren't usual tcp_client things - maybe mark which commands are only used when reversed?

File: src/riak_repl_tcp_client.erl [hunk 2]
```diff
@@ -303,47 +327,64 @@ handle_peerinfo(#state{sitename=SiteName, socket=Socket} = State, TheirPeerInfo,
...
+
+                    ServerStrats = proplists:get_value(fullsync_strategies, Capability,
+                        [?LEGACY_STRATEGY]),
+                    ClientStrats = app_helper:get_env(riak_repl, fullsync_strategies,
+                        [?LEGACY_STRATEGY]),
+                    Strategy = riak_repl_util:choose_strategy(ServerStrats, ClientStrats),
+                    StratMod = riak_repl_util:strategy_module(Strategy, client),
+                    lager:info("Using fullsync strategy ~p with site ~p.", [StratMod,
+                            State#state.sitename]),
+                    lager:notice("making workdir"),
```
Comment (jonmeredith) [inline src/riak_repl_tcp_client.erl]: Sneaky log message crept in.

**Comment by jonmeredith** [comment]
I tried modifying the repl2x2 testin basho_expect and managed to get a couple of crashes.  Looks like tcp_client received a {ref(), keylist_built} message.

b_e branch pushed as jdm-rev-repl2x2

```

2012-04-27 13:40:44 =SUPERVISOR REPORT====
     Supervisor: {local,riak_repl_client_sup}
     Context:    child_terminated
     Reason:     normal
     Offender:   [{pid,<0.1981.0>},{name,"master"},{mfargs,{riak_repl_tcp_client,start_link,["master"]}},{restart_type,permanent},{shutdown,brutal_kill},{child_type,worker}]

2012-04-27 13:40:45 =ERROR REPORT====
** State machine <0.2502.0> terminating 
** Last event in was {#Ref<0.0.0.14100>,keylist_built}
** When State == build_keylist
**      Data  == {state,"master",#Port<0.8995>,"/var/lib/riak/riak_repl/work/18709816/master-192.168.169.100:57633-192.168.169.115:5558",{riak_client,'riak@192.168.169.100',undefined},<0.2525.0>,#Ref<0.0.0.14109>,"/var/lib/riak/riak_repl/work/18709816/master-192.168.169.100:57633-192.168.169.115:5558/479555224749202520035584085735030365824602865664.ours.sterm",undefined,"/var/lib/riak/riak_repl/work/18709816/master-192.168.169.100:57633-192.168.169.115:5558/479555224749202520035584085735030365824602865664.theirs.sterm",undefined,479555224749202520035584085735030365824602865664,undefined,undefined,{1335,555645,306370},{1335,555645,304293},<0.2503.0>,true,100}
** Reason for termination = 
** {function_clause,[{riak_repl_keylist_server,build_keylist,[{#Ref<0.0.0.14100>,keylist_built},{state,"master",#Port<0.8995>,"/var/lib/riak/riak_repl/work/18709816/master-192.168.169.100:57633-192.168.169.115:5558",{riak_client,'riak@192.168.169.100',undefined},<0.2525.0>,#Ref<0.0.0.14109>,"/var/lib/riak/riak_repl/work/18709816/master-192.168.169.100:57633-192.168.169.115:5558/479555224749202520035584085735030365824602865664.ours.sterm",undefined,"/var/lib/riak/riak_repl/work/18709816/master-192.168.169.100:57633-192.168.169.115:5558/479555224749202520035584085735030365824602865664.theirs.sterm",undefined,479555224749202520035584085735030365824602865664,undefined,undefined,{1335,555645,306370},{1335,555645,304293},<0.2503.0>,true,100}]},{gen_fsm,handle_msg,7},{proc_lib,init_p_do_apply,3}]}
```

**Comment by jonmeredith** [comment]
Logs here for a while http://dl.dropbox.com/u/2596739/revconfail.tar.gz

**Comment by Vagabond** [comment]
I'm pretty sure this is a bug present in the current protocol, and that you're only seeing it with the connection reversal stuff enabled by coincidence.

I think what is happening is that the 'continue' message sent in riak_repl_tcp_server:wait_for_partition is not being consumed because the client is sending a `skip_partition` message that ends up before it in the message queue. If you look at the logs, you can see the same partition being synced 5 times, because those continue messages aren't being discarded when the `skip_partition` message is received.

**Comment by jonmeredith** [comment]
Couple of new issues, first the realtime replication hook is installed on the 'server' and not the 'client' so the objects written on the client are not sent to the server.

On the 'server' leader (node that listens for connection)

```

riak_core_bucket:get_bucket(<<"systest">>).
[{name,<<"systest">>},
 {allow_mult,false},
 {basic_quorum,false},
 {big_vclock,50},
 {chash_keyfun,{riak_core_util,chash_std_keyfun}},
 {dw,quorum},
 {last_write_wins,false},
 {linkfun,{modfun,riak_kv_wm_link_walker,mapreduce_linkfun}},
 {n_val,3},
 {notfound_ok,true},
 {old_vclock,86400},
 {postcommit,[{struct,[{<<"mod">>,<<"riak_repl_leader">>},
                       {<<"fun">>,<<"postcommit">>}]}]},
 {pr,0},
 {precommit,[]},
 {pw,0},
 {r,quorum},
 {repl,true},
 {rw,quorum},
 {small_vclock,50},
 {w,quorum},
 {young_vclock,20}]
```

On the 'client' leader (node the connects to remote socket)

```
riak_core_bucket:get_bucket(<<"systest">>).
[{name,<<"systest">>},
 {allow_mult,false},
 {basic_quorum,false},
 {big_vclock,50},
 {chash_keyfun,{riak_core_util,chash_std_keyfun}},
 {dw,quorum},
 {last_write_wins,false},
 {linkfun,{modfun,riak_kv_wm_link_walker,mapreduce_linkfun}},
 {n_val,3},
 {notfound_ok,true},
 {old_vclock,86400},
 {postcommit,[]},
 {pr,0},
 {precommit,[]},
 {pw,0},
 {r,quorum},
 {rw,quorum},
 {small_vclock,50},
 {w,quorum},
 {young_vclock,20}]
```

**Comment by jonmeredith** [comment]
Second issue, fullsync does not exchange data.

I wrote a thousand objects to the 'client' cluster while the 'server' cluster was down and reconnected.

Output on the server (machine that was listening for connections)

```
(riak@192.168.169.122)3> 13:57:35.061 [info] Shuffling partition list using seed {1336,420655,61887}
13:57:35.062 [info] Full-sync with site "c2" starting; 64 partitions.
13:57:35.062 [info] Full-sync with site "c2"; starting fullsync for 1073290264914881830555831049026020342559825461248
13:57:35.062 [info] Full-sync with site "c2"; building keylist for 1073290264914881830555831049026020342559825461248, 63 remain
13:57:35.099 [info] Sorting keylist "/var/lib/riak/riak_repl/work/27503143/c2-192.168.169.122:9011-192.168.169.115:38048/1073290264914881830555831049026020342559825461248.ours.sterm"
13:57:35.100 [info] Sorted /var/lib/riak/riak_repl/work/27503143/c2-192.168.169.122:9011-192.168.169.115:38048/1073290264914881830555831049026020342559825461248.ours.sterm in 0.00 seconds
13:57:35.100 [info] Full-sync with site "c2"; built keylist for 1073290264914881830555831049026020342559825461248, (built in 0.03 secs)
13:57:35.100 [info] Full-sync for "c2"; sending keylist for 1073290264914881830555831049026020342559825461248
13:57:35.100 [info] Full-sync with site "c2"; sent keylist for 1073290264914881830555831049026020342559825461248 (sent in 0.0 secs)
13:57:35.101 [info] Full-sync with site "c2"; exhanging differences for 1073290264914881830555831049026020342559825461248
13:57:35.105 [info] Full-sync with site "c2"; differences exchanged for 1073290264914881830555831049026020342559825461248 (done in 0.0 secs)
13:57:35.105 [info] Full-sync with site "c2"; full-sync for partition 1073290264914881830555831049026020342559825461248 complete (done in 0.04 secs)
```

Output on the client (machine that opened TCP connection)

```
13:57:35.055 [info] Full-sync with site "c2"; doing fullsync for 1073290264914881830555831049026020342559825461248
13:57:35.056 [info] Full-sync with site "c2"; building keylist for 1073290264914881830555831049026020342559825461248
13:57:35.059 [info] Sorting keylist "/var/lib/riak/riak_repl/work/44469249/c2-192.168.169.115:38048-192.168.169.122:9011/1073290264914881830555831049026020342559825461248.ours.sterm"
13:57:35.060 [info] Sorted /var/lib/riak/riak_repl/work/44469249/c2-192.168.169.115:38048-192.168.169.122:9011/1073290264914881830555831049026020342559825461248.ours.sterm in 0.00 seconds
13:57:35.060 [info] Full-sync with site "c2"; built keylist for 1073290264914881830555831049026020342559825461248 (built in 0.0 secs) 
13:57:35.094 [info] Full-sync with site "c2"; received keylist for 1073290264914881830555831049026020342559825461248 (received in 0.03 secs)
13:57:35.094 [info] Full-sync with site "c2"; calculating differences for 1073290264914881830555831049026020342559825461248
13:57:35.095 [info] Partition 1073290264914881830555831049026020342559825461248: 43 remote / 0 local: 43 missing, 0 differences.
13:57:35.098 [info] Full-sync with site "c2"; differences exchanged for partition 1073290264914881830555831049026020342559825461248 (done in 0.0 secs)
13:57:35.098 [info] Full-sync with site "c2"; fullsync for 1073290264914881830555831049026020342559825461248 complete (completed in 0.04 secs)
```

interestingly it normally reports things in a confusing way...  this is with repl set up with invert_connections, false on the server node.
I would expect the local object count to be 53 and the remote 0, 53 missing, 0 differences.

```

(riak@192.168.169.100)3> 14:01:27.565 [info] Using fullsync strategy riak_repl_keylist_server.
14:01:27.571 [info] Full-sync on connect
14:01:27.598 [info] Full-sync with site "c2"; doing fullsync for 1370157784997721485815954530671515330927436759040
14:01:27.598 [info] Full-sync with site "c2"; building keylist for 1370157784997721485815954530671515330927436759040
14:01:27.604 [info] Sorting keylist "/var/lib/riak/riak_repl/work/10621062/c2-192.168.169.100:9011-192.168.169.115:47594/1370157784997721485815954530671515330927436759040.ours.sterm"
14:01:27.608 [info] Sorted /var/lib/riak/riak_repl/work/10621062/c2-192.168.169.100:9011-192.168.169.115:47594/1370157784997721485815954530671515330927436759040.ours.sterm in 0.00 seconds
14:01:27.608 [info] Full-sync with site "c2"; built keylist for 1370157784997721485815954530671515330927436759040 (built in 0.0 secs) 
14:01:27.619 [info] Full-sync with site "c2"; received keylist for 1370157784997721485815954530671515330927436759040 (received in 0.01 secs)
14:01:27.620 [info] Full-sync with site "c2"; calculating differences for 1370157784997721485815954530671515330927436759040
14:01:27.621 [info] Partition 1370157784997721485815954530671515330927436759040: 53 remote / 0 local: 53 missing, 0 differences.
14:01:27.626 [info] Full-sync with site "c2"; differences exchanged for partition 1370157784997721485815954530671515330927436759040 (done in 0.0 secs)
14:01:27.627 [info] Full-sync with site "c2"; fullsync for 1370157784997721485815954530671515330927436759040 complete (completed in 0.02 secs)
```

File: src/riak_repl_tcp_client.erl (hunk 2 above)
Comment (jonmeredith) [inline src/riak_repl_tcp_client.erl]: Sneaky log message still here.

**Comment by jonmeredith** [comment]
+1 merge.

Would be nice to remove that log message about making directories.

Added basho_bench jdm-rev-repl2x2 branch that tests the inverse connection.

**Pull request merged**